File: learn/stack-and-queue/stack-array/stack_array.go


type StackArray struct {
	arr      []int
	capacity int
}

func NewStackArray(size int) *StackArray {
	return &StackArray{
		arr:      make([]int, 0, size),
		capacity: size,
	}
}

func (s *StackArray) Push(value int) {
	if len(s.arr) < s.capacity {
		s.arr = append(s.arr, value)
	}
}

func (s *StackArray) Pop() int {
	if s.IsEmpty() {
		return -1
	}
	val := s.arr[len(s.arr)-1]
	s.arr = s.arr[:len(s.arr)-1]
	return val
}

func (s *StackArray) Peek() int {
	if s.IsEmpty() {
		return -1
	}
	return s.arr[len(s.arr)-1]
}

func (s *StackArray) IsEmpty() bool {
	return len(s.arr) == 0
}

func (s *StackArray) Size() int {
	return len(s.arr)
}

File: learn/stack-and-queue/stack-array/stack_array.rs
pub struct StackArray {
    arr: Vec<i32>,
    capacity: usize,
}

impl StackArray {
    pub fn new(capacity: usize) -> Self {
        StackArray {
            arr: Vec::with_capacity(capacity),
            capacity,
        }
    }

    pub fn push(&mut self, value: i32) {
        if self.arr.len() < self.capacity {
            self.arr.push(value);
        }
    }

    pub fn pop(&mut self) -> Option<i32> {
        self.arr.pop()
    }

    pub fn peek(&self) -> Option<&i32> {
        self.arr.last()
    }

    pub fn is_empty(&self) -> bool {
        self.arr.is_empty()
    }

    pub fn size(&self) -> usize {
        self.arr.len()
    }
}
